src/components/application_manager/test/rpc_passing_handler_test.cc

```diff
@@ -0,0 +1,447 @@
+                                    am::commands::Command::SOURCE_MOBILE));
+  }
+  bool result = rpc_passing_handler_->RPCPassThrough(unsupported_response);
+  EXPECT_EQ(result, true);
+}
+
+TEST_F(RPCPassingHandlerTest,
+       RPCPassingTest_UNSUPPORTED_REQUEST_ForwardToMobile) {
+  smart_objects::SmartObject unsupported_response = CreatePassThroughResponse(
+      kConnectionKey_NAV_ASP, kCorrelationId, "UNSUPPORTED_REQUEST", false);
+  smart_objects::SmartObject forwarded_request =
+      CreatePassThroughRequest(kConnectionKey_MEDIA_ASP, kCorrelationId);
+
+  app_services_.push_back(CreateAppService(
+      kConnectionKey_NAV_ASP, "Navigation service", "NAVIGATION"));
+  app_services_.push_back(
+      CreateAppService(kConnectionKey_MEDIA_ASP, "Media service", "MEDIA"));
+
+  PassthroughParams request_params{
+      .connection_key = kConnectionKey_ASC,
+      .correlation_id = kCorrelationId,
+      .message = CreatePassThroughRequest(kConnectionKey_ASC, kCorrelationId),
+      .forwarded_message =
+          CreatePassThroughRequest(kConnectionKey_NAV_ASP, kCorrelationId),
+  };
+
+  SendRequestToASP(request_params);
+
+  {
+    InSequence dummy;
+    // Call RPCPassThrough with response smart object
+    // Will cycle to next compatible active app service
+    EXPECT_CALL(mock_app_manager_, get_settings());
+    EXPECT_CALL(mock_app_manager_settings_, rpc_pass_through_timeout());
+    EXPECT_CALL(mock_app_manager_, GetRPCService());
+    EXPECT_CALL(mock_rpc_service_,
+                SendMessageToMobile(Pointee(forwarded_request), false));
+  }
+  bool result = rpc_passing_handler_->RPCPassThrough(unsupported_response);
+  EXPECT_EQ(result, true);
+}
+
+TEST_F(RPCPassingHandlerTest, RPCPassingTest_REQUEST_Timeout) {
+  app_services_.push_back(CreateAppService(
+      kConnectionKey_NAV_ASP, "Navigation service", "NAVIGATION"));
+
+  PassthroughParams request_params{
+      .connection_key = kConnectionKey_ASC,
+      .correlation_id = kCorrelationId,
+      .message = CreatePassThroughRequest(kConnectionKey_ASC, kCorrelationId),
+      .forwarded_message =
+          CreatePassThroughRequest(kConnectionKey_NAV_ASP, kCorrelationId),
+  };
+
+  SendRequestToASP(request_params, 1);
+
+  {
+    InSequence dummy;
+    // Request timeout will trigger perform next request
+    // Will cycle to core (no other app services in list)
+    EXPECT_CALL(mock_app_manager_, application(kConnectionKey_ASC));
+    EXPECT_CALL(*mock_app_ptr_, msg_version());
+    EXPECT_CALL(mock_app_manager_, GetRPCHandler());
+    EXPECT_CALL(mock_rpc_handler_,
+                ValidateRpcSO(request_params.message, _, _, false))
+        .WillOnce(Return(true));
+    EXPECT_CALL(mock_app_manager_, GetRPCService());
+    EXPECT_CALL(mock_rpc_service_,
+                ManageMobileCommand(Pointee(request_params.message),
+                                    am::commands::Command::SOURCE_MOBILE));
+  }
+  sleep(1);
```

Comment:
Look at `TestASyncWaiter` for asynchronous expectations

Reply:
Resolved in https://github.com/smartdevicelink/sdl_core/pull/2912/commits/d194dfd9e425d0539a1c4a487eaf5dd389cd623b